src/authentication/authentication-hapi-plugin.ts

```diff
@@ -0,0 +1,154 @@
+
+import * as Boom from 'boom';
+import * as auth from 'hapi-auth-jwt2';
+import { inject, injectable } from 'inversify';
+import * as Joi from 'joi';
+import * as qs from 'querystring';
+
+import * as Hapi from '../server/hapi';
+import { HapiPlugin } from '../server/hapi-register';
+import { Group, User } from '../shared/gitlab';
+import { GitlabClient } from '../shared/gitlab-client';
+import { Logger, loggerInjectSymbol } from '../shared/logger';
+import {default as AuthenticationModule} from './authentication-module';
+import { AccessToken, authServerBaseUrlInjectSymbol, jwtOptionsInjectSymbol } from './types';
+
+interface UserId {
+  id: string;
+  idp: string;
+}
+
+@injectable()
+class AuthenticationHapiPlugin extends HapiPlugin {
+
+  public static injectSymbol = Symbol('authentication-hapi-plugin');
+  private readonly authenticationModule: AuthenticationModule;
+  private readonly authServerBaseUrl: string;
+  private readonly logger: Logger;
+  private readonly gitlab: GitlabClient;
+  private readonly options: auth.JWTStrategyOptions;
+  private teamTokenClaimKey: 'https://team_token' = 'https://team_token';
+  private subEmailClaimKey: 'https://sub_email' = 'https://sub_email';
+
+  constructor(
+    @inject(AuthenticationModule.injectSymbol) authenticationModule: AuthenticationModule,
+    @inject(authServerBaseUrlInjectSymbol) authServerBaseUrl: string,
+    @inject(loggerInjectSymbol) logger: Logger,
+    @inject(GitlabClient.injectSymbol) gitlab: GitlabClient,
+    @inject(jwtOptionsInjectSymbol) jwtOptions: auth.JWTStrategyOptions,
+    ) {
```

Comment:
I noticed that the code is formatted in this way throughout the project, so this comment isn't specific to this PR, but it's something that (to me) improves code legibility: similarly to how e.g. `super` is formatted below, I like to indent closing parenthesis on the same line as where the calling function is. This lets you very easily see where the list of parameters ends and where the function starts.